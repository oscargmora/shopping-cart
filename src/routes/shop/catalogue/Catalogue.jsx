import { useEffect, useState } from "react";
import PropTypes from 'prop-types';
import Header from "../../../header/Header";
import { getRequestWithNativeFetch } from "./FetchData";
import { Counter } from "../counter/Counter";
import catalogueStyles from './catalogue.module.css';

const Catalogue = ({ addToBasket, basket }) => {
    const [items, setItems] = useState(null);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);

    useEffect(() => {
        const fetchItems = async() => {
            try {
                const itemsData = await getRequestWithNativeFetch(
                    'https://fakestoreapi.com/products'
                );
                setItems(itemsData);
                setError(null);
            } catch (err) {
                setError(err.message);
                setItems(null);
            } finally {
                setLoading(false);
            }
        };

        fetchItems();
    }, []);

    return (
        <>
            <Header />

            <div className={catalogueStyles.ulContainer}>
                {loading && (
                    <div>Loading Catalogue...</div>
                )}

                {error && <div>{error}</div>}

                <ul className={catalogueStyles.ul}>
                    {
                        items && items.map(({id, title, image, price}) => (
                            <li className={catalogueStyles.item} key={id}>
                                <img className={catalogueStyles.img} src={image} />
                                <p className={catalogueStyles.title}>{title}</p>
                                <p>${price}</p>
                                <Counter dataId={id} basket={basket} />
                                <button onClick={(e) => addToBasket(id, Number(e.target.previousSibling.firstChild.nextSibling.innerText), title, image, price)}>Add To Cart</button>
                            </li>
                        ))
                    }
                </ul>
            </div>
        </>
    )
}

Catalogue.propTypes = {
    addToBasket: PropTypes.func.isRequired,
    basket: PropTypes.object.isRequired,
}

export default Catalogue;