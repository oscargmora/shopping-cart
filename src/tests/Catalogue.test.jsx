import {render, screen} from '@testing-library/react'
import { describe, it, expect } from 'vitest';
import Catalogue from '../routes/shop/catalogue/Catalogue';
import { BrowserRouter } from 'react-router-dom';

describe('Catalogue', () => {
    it('renders "Loading Catalogue..."', () => {
        render(<BrowserRouter><Catalogue /></BrowserRouter>);
        expect(screen.getByText('Loading Catalogue...')).toBeInTheDocument();
    });
});