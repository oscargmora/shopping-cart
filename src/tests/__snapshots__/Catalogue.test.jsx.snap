// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart API Call > renders App component 1`] = `
<div>
  <div
    class="_headerContainer_aaeeca"
  >
    <header
      class="_header_aaeeca"
    >
      <a
        class="_link_aaeeca"
        href="/"
      >
        <h1>
          Fake Store
        </h1>
      </a>
      <div
        class="_headerMain_aaeeca"
      >
        <a
          class="_link_aaeeca"
          href="/"
        >
          Home
        </a>
        <a
          class="_link_aaeeca"
          href="/catalogue"
        >
          Catalogue
        </a>
      </div>
      <a
        href="/cart"
      >
        <img
          class="_shoppingCart_aaeeca"
          src="/src/assets/shopping-cart.png"
        />
      </a>
    </header>
    <hr />
  </div>
  <div
    class="_ulContainer_e1bb8e"
  >
    <p
      class="_loading_e1bb8e"
    >
      Loading Catalogue...
    </p>
    <ul
      class="_ul_e1bb8e"
    />
  </div>
</div>
`;
